commit d5e722168a84
Author: Dragana Damjanovic <dd.mozilla@gmail.com>
Date:   Wed Dec 20 16:13:36 2017 +0100

    Bug 1422545 - Do not close connection between a httpChannelChild and its httpChannelParent if we need to divert to parent. r=mayhemer, a=jcristau
    
    If UnknowDecoder is involved and the received content is short we will know whether we need to divert to parent only after OnStartRequest of the listener chain is called. Therefore do not do cleanup if we detect diversion.
    
    --HG--
    extra : source : e87b07c9db7c302237fa0adbb68c0f31b1840191
---
 netwerk/protocol/http/HttpChannelChild.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git netwerk/protocol/http/HttpChannelChild.cpp netwerk/protocol/http/HttpChannelChild.cpp
index 18b9fe2de1db..8fc7746a2a42 100644
--- netwerk/protocol/http/HttpChannelChild.cpp
+++ netwerk/protocol/http/HttpChannelChild.cpp
@@ -1100,6 +1100,19 @@ HttpChannelChild::OnStopRequest(const nsresult& channelStatus,
     // DoOnStopRequest() calls ReleaseListeners()
   }
 
+  // If unknownDecoder is involved and the received content is short we will
+  // know whether we need to divert to parent only after OnStopRequest of the
+  // listeners chain is called in DoOnStopRequest. At that moment
+  // unknownDecoder will call OnStartRequest of the real listeners of the
+  // channel including the OnStopRequest of UrlLoader which decides whether we
+  // need to divert to parent.
+  // If we are diverting to parent we should not do a cleanup.
+  if (mDivertingToParent) {
+    LOG(("HttpChannelChild::OnStopRequest  - We are diverting to parent, "
+         "postpone cleaning up."));
+    return;
+  }
+
   CleanupBackgroundChannel();
 
   // If there is a possibility we might want to write alt data to the cache
